### PR TITLE
EVAL-02: basic work on compiler API wrapper

### PR DESCRIPTION
**Pull request by jyoo980**
I've also included the example script from the TypeScript compiler API page.

Changing some of the values in that file with the values of our function calls should be a quick + dirty way for us to test our code other than unit testing.

### CODE REVIEW

**Review by scveloso — APPROVED** [review]
i like it i like it

File: src/ast/DirDecl.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Represents the directories a TypeScript project may have.
+ *
+ * e.g. dirs [src/model/model, test/util]
```
Comment (uslava77) [inline src/ast/DirDecl.ts]: dirs are different now (nested version)

File: src/ast/FieldDecl.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import {VarList} from "./VarList";
+
+/**
+ * Represents a field declaration in the TypeScript DSL.
+ *
+ * e.g. fields [bar string, baz boolean] as public
+ */
+export class FieldDecl {
```
Comment (uslava77) [inline src/ast/FieldDecl.ts]: should these extend an abstract class Node?

File: src/ast/DirDecl.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ */
+export class DirDecl {
```
Comment (uslava77) [inline src/ast/DirDecl.ts]: should these extend an abstract class Node?
and parse evaluate functions throughout?

File: src/ast/ModuleDecl.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Represents the dependencies a TypeScript project may have. Development dependencies by default.
+ *
+ * e.g. modules [@types/chai: 4.0.8, chai: 4.1.2, mocha: 4.1.0]
+ */
+export class ModuleDecl {
```
Comment (uslava77) [inline src/ast/ModuleDecl.ts]: should these extend an abstract class Node?

File: src/ast/ModuleDecl.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Represents the dependencies a TypeScript project may have. Development dependencies by default.
+ *
+ * e.g. modules [@types/chai: 4.0.8, chai: 4.1.2, mocha: 4.1.0]
```
Comment (uslava77) [inline src/ast/ModuleDecl.ts]: should have quotations around modules

File: src/ast/FieldDecl.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import {VarList} from "./VarList";
+
+/**
+ * Represents a field declaration in the TypeScript DSL.
+ *
+ * e.g. fields [bar string, baz boolean] as public
```
Comment (uslava77) [inline src/ast/FieldDecl.ts]: this is different now

```
fields public [string bar, boolean baz]
        generate getters
        generate setters
```

File: src/ast/DirDecl.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+    directories: string[];
+
+    constructor(directories: string[]) {
+        this.directories = directories;
```
Comment (uslava77) [inline src/ast/DirDecl.ts]: should initialization happen at parse()?

Comment (uslava77) [reply src/ast/DirDecl.ts]: this comment applies for all constructors in this PR

File: src/ast/VarList.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Represents a list of fields and their corresponding type. Used in declaring fields.
+ *
+ * e.g. [foo string, bar boolean, baz IInsightFacade]
```
Comment (uslava77) [inline src/ast/VarList.ts]: [string foo, boolean bar, IInsightFacade baz]

**Review by uslava77 — COMMENTED** [review]
Haven't finished but here are some comments, will look at the other files tomorrow

**Comment by jyoo980** [comment]
@uslava77 

Addressing your comments - I made these before we finalized the EBNF so I'll make the necessary changes and push again!

File: src/ast/AstNode.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+export abstract class AstNode {
+
```
Comment (kiyomih) [inline src/ast/AstNode.ts:2]: nit: would prefer if this is an interface unless we have a specific reason for making this an abstract class (like we're planning to include some implementation here). 
Reasons: we can't easily extend multiple classes in typescript (and there may be a case we want to do this at some point?), and if we're just implementing this abstract class, then it's just treated as an interface anyways.

**Review by kiyomih — APPROVED** [review]
Looks great! 🔥💯🙌🏻 Just left a small nit comment, but would be good with this being merged as is.

File: src/ast/AstNode.ts (hunk 9 above)
Comment (jyoo980) [reply src/ast/AstNode.ts:2]: @kiyomih thanks for your feedback!

I agree that usually, behaviour is better specified as an interface. I was thinking that we have some of the common classes, e.g. `Tokenizer`, `FileSystem`, live inside the abstract `AstNode` class, much like how Elisa implemented tinyDot in class. 

I'll wait for your response before I merge; architectural designs like these are important, thanks for bringing it up!

Comment (kiyomih) [reply src/ast/AstNode.ts:2]: That's a fair point! Everything looks good to me in that case.

File: src/ast/symbols/TypeTable.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        return typeNodePair.pair[1];
+    }
+
+    /**
+     * Returns true iff the strings passed in are valid types
+     * Could be useful for validating field declarations.
+     * @param candidates
+     */
+    public areValidTypes(candidates: string[]): boolean {
+        return candidates.reduce((acc, val) => {
```
Comment (daviidli) [inline src/ast/symbols/TypeTable.ts]: Is this returning true iff **all** candidates are valid types?
If so, we can do this instead: 
```typescript
candidates.every((candidate) => this.isValidType(candidate));
```
As is, it just returns if the last candidate is valid or not

**Review by daviidli — APPROVED** [review]
Looks good to me, just a small comment.

File: src/ast/symbols/TypeTable.ts (hunk 10 above)
Comment (jyoo980) [reply src/ast/symbols/TypeTable.ts]: You're right - changing!

**Pull request merged**